### ISSUE TITLE
Dyninst-10.0.0 undeclared EM_AARCH64

### ISSUE THREAD

**Issue by vysocky** [dyninst/dyninst#503]
When building Dyninst-10.0.0 on our x86 cluster the compilation fails on undeclared EM_AARCH64.
```
> [ 22%] Building CXX object symtabAPI/CMakeFiles/symtabAPI.dir/src/Object-elf.C.o
> /home/vys0053/Software/Dyninst-10.0.0/symtabAPI/src/Object-elf.C: In function ‘Dyninst::SymtabAPI::Region::RegionType getRelTypeByElfMachine(Dyninst::Elf_X*)’:
> /home/vys0053/Software/Dyninst-10.0.0/symtabAPI/src/Object-elf.C:277:14: error: ‘EM_AARCH64’ was not declared in this scope
>          case EM_AARCH64:
>               ^~~~~~~~~~
```
Using GCC/6.3.0, Boost/1.66.0. Compilation of Dyninst-9.3.2 works fine.

[See complete build output.](https://pastebin.com/rswJvDNF)

**Comment by mxz297** [comment]
@sashanicolas Are you able to reproduce this error?

**Comment by vysocky** [comment]
On Ubuntu the compilation was fine. Btw we are going to upgrade CentoOS 6.9 to CentoOS 7 next week, so the environment may change at my side.

**Comment by mxz297** [comment]
A quick potential fix is to add the following three lines to symtabAPI/src/Object-elf.C

```
#if !defined(EM_AARCH64)
#define EM_AARCH64 183
#endif
```

Give it a try and see what we got.

**Comment by vysocky** [comment]
Then there are another missing variables:
```
[ 23%] Building CXX object symtabAPI/CMakeFiles/symtabAPI.dir/src/emitElfStatic.C.o
/home/vys0053/Software/Dyninst-10.0.0/symtabAPI/src/emitElfStatic.C: In static member function ‘static bool Dyninst::SymtabAPI::emitElfUtils::updateRelocation(Dyninst::SymtabAPI::Symtab*, Dyninst::SymtabAPI::relocationEntry&, int)’:
/home/vys0053/Software/Dyninst-10.0.0/symtabAPI/src/emitElfStatic.C:1665:18: error: ‘R_AARCH64_RELATIVE’ was not declared in this scope
             case R_AARCH64_RELATIVE:
                  ^
/home/vys0053/Software/Dyninst-10.0.0/symtabAPI/src/emitElfStatic.C:1669:18: error: ‘R_AARCH64_JUMP_SLOT’ was not declared in this scope
             case R_AARCH64_JUMP_SLOT:
                  ^
/home/vys0053/Software/Dyninst-10.0.0/symtabAPI/src/emitElfStatic.C:1678:18: error: ‘R_AARCH64_GLOB_DAT’ was not declared in this scope
             case R_AARCH64_GLOB_DAT:
```

**Comment by mxz297** [comment]
@vysocky These missing declarations are all about ARM. On ubuntu, even if the architecture is x86, the system header files define those ARM related macros (and macros for other architectures). It looks like that on CentOS, the header files installed on x86 will not include the macros for ARM, causing these compilation errors. 

I will ask around and create a CentOS environment to see how to fix it.

**Comment by mxz297** [comment]
@vysocky Meanwhile, since the missing declarations are about ARM binaries, you can work around them by just commenting out those missing variables.

**Comment by vysocky** [comment]
Changes only in these two files have been necessary to compile the Dyninst-10.0.0 successfully
```
symtabAPI/src/Object-elf.C - EM_AARCH64
symtabAPI/src/emitElfStatic.C - R_AARCH64_RELATIVE, R_AARCH64_JUMP_SLOT, R_AARCH64_GLOB_DAT
```

**Comment by mxz297** [comment]
@vysocky  Thanks!

@sashanicolas I think we should refactor the involved code from the architecture independent source file (emitElfStatic.C) to architecture dependent source file (emitElfStatic-aarch64.C).

**Comment by sashanicolas** [comment]
fixed by commit 58985e9c66d0bfb56a8e26442c07c67564968837
